[Twig] Add NotificationEmail

fabpot opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | n/a
| License       | MIT
| Doc PR        | -

This PR is an extract of the new Notifier component. It's a default template to send standardized emails with the Mailer component, which can be used independently of the Notifier component.

Such emails look like the following:

<img width="618" alt="image" src="https://user-images.githubusercontent.com/47313/65018883-694cb780-d92a-11e9-940a-875ea68f9e5d.png">

More info on SpeakerDeck (be warned that names have change since my presentation): https://speakerdeck.com/fabpot/symfony-notifier?slide=7

It requires Twig 1.12 which should be released later this week.

Usage example:

```php
$email = (new NotificationEmail())
    ->from('fabien@example.com')
    ->to('fabien@example.org')
    ->subject('My first notification email via Symfony')
    ->markdown(<<<EOF
There is a **problem** on your website, you should investigate it right now.
Or just wait, the problem might solves itself automatically, we never know.
EOF
    )
    ->action('More info?', 'https://example.com/')
    ->importance('high')
    //->exception(new \LogicException('That does not work at all...'))
;
```

Instead of `markdown()`, you can also use `content()` for simple emails.

Note that you can use Inky tags in the content:

```php
$email = (new NotificationEmail())
    ->from('fabien@example.com')
    ->to('fabien@example.org')
    ->subject('My first notification email via Symfony')
    ->markdown(<<<EOF
There is a **problem** on your website, you should investigate it right now.
Or just wait, the problem might solves itself automatically, we never know.

Some Title
==========

<center>
    <button href="https://example.com/">Go?</button>
</center>

EOF
);
```

There is also the concept of a theme. By default, it uses the `default` theme,  which is an alias for the `zurb_2` theme.

You can use `setTheme()` to override the theme for a given instance, or override the themes globally via the following config in `twig.yaml`:

```yaml
twig:
    paths:
        templates/email: email
```

Then, create `templates/email/default/notification/body.html.twig` and `templates/email/default/notification/body.txt.twig`. Extends the existing template via `{% extends "@!email/default/notification/body.html.twig" %}` (note  the `!`).

fabpot committed:
[Twig] Add NotificationEmail

fabpot commented:
Usage example:

```php
$email = (new NotificationEmail())
    ->from('fabien@example.com')
    ->to('fabien@example.org')
    ->subject('My first notification email via Symfony')
    ->markdown()
    ->content(<<<EOF
There is a **problem** on your website, you should investigate it right now.
Or just wait, the problem might solves itself automatically, we never know.
EOF
    )
    ->action('More info?', 'https://example.com/')
    ->importance('high')
    //->exception(new \LogicException('That does not work at all...'))
;
```

lyrixx commented on src/Symfony/Bridge/Twig/Mime/NotificationEmail.php:
```diff
@@ -0,0 +1,202 @@
+    /**
+     * @return $this
+     */
+    public function markdown()
```
IMHO this should accept a string (the markdown content)

fabpot replied:
Done

lyrixx commented on src/Symfony/Bridge/Twig/Mime/NotificationEmail.php:
```diff
@@ -0,0 +1,202 @@
+    public function exception($exception)
+    {
+        $exceptionAsString = '';
+        if (!class_exists(FlattenException::class)) {
```
It's better for readability to start with the positive condition

rbaarsma reviewed (commented):
Shouldn't there be automated tests before merging this change?

yceruto commented on src/Symfony/Bundle/TwigBundle/DependencyInjection/Compiler/ExtensionPass.php:
```diff
@@ -40,11 +40,17 @@ public function process(ContainerBuilder $container)
+        $viewDir = \dirname(new \ReflectionClass('Symfony\Bridge\Twig\Extension\FormExtension'), 2).'/Resources/views';
+
+        if ($container->has('mailer')) {
+            $emailPath = $viewDir.'/Email';
```
This new path should also be added to the `twig.template_iterator` service to make it cacheable on `cache:clear`.

fabpot replied:
done

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -1585,6 +1585,13 @@ private function addMailerSection(ArrayNodeDefinition $rootNode)
                                 ->end()
                             ->end()
                         ->end()
+                        ->arrayNode('notification')
```
this seems unrelated to this PR, but to the notifier component instead, isn't it ?

fabpot replied:
It is related. It allows to change the default templates for the `NotificationEmail` class.

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -2006,6 +2007,8 @@ private function registerMailerConfiguration(array $config, ContainerBuilder $co
         $envelopeListener->setArgument(0, $sender);
         $envelopeListener->setArgument(1, $recipients);
+
+        NotificationEmail::setDefaultTemplates($config['notification']['templates'][0] ?? null, $config['notification']['templates'][1] ?? null);
```
This looks wrong to me. This static call will be done only at compile time, and so won't be applied for any request using a cached container. This needs to be moved to `boot()` of the bundle (or dropped)

fabpot replied:
Oops, fixed.

fabpot commented on src/Symfony/Bridge/Twig/Mime/NotificationEmail.php:
```diff
@@ -0,0 +1,220 @@
+    public const URGENT = 'urgent';
+    public const HIGH = 'high';
+    public const MEDIUM = 'medium';
+    public const LOW = 'low';
```
I went back and forth about this one, but having a prefix is probably more consistent with the rest of the codebase. Renamed.

fabpot commented on src/Symfony/Bridge/Twig/Mime/NotificationEmail.php:
```diff
@@ -0,0 +1,220 @@
+                return self::PRIORITY_NORMAL;
+            case self::LOW:
+                return self::PRIORITY_LOW;
+            default:
```
Exactly, just defensive programming, but this should indeed never happen for now.

stof replied:
then, I suggest using `self::PRIORITY_LOW` for this defensive path, so that this path does not unlock using a lowest priority than support code paths (and then having people wondering how they can trigger that code path without bugs elsewhere)

fabpot commented on src/Symfony/Bridge/Twig/Resources/views/Email/zurb.css:
```diff
@@ -0,0 +1,1667 @@
+/*
+ * Copyright (c) 2017 ZURB, inc. -- MIT License
+ *
+ * https://raw.githubusercontent.com/foundation/foundation-emails/develop/dist/foundation-emails.css
```
changed the linked and renamed the file to `zurb_foundation_emails_2.css`.

stof replied:
when switching to themes, the naming went back to `zurb` only.

fabpot commented on src/Symfony/Bridge/Twig/Resources/views/Email/notification_body_markdown.html.twig:
```diff
@@ -0,0 +1,3 @@
+{% filter markdown_to_html %}
+    {{ raw ? content|raw : content }}
```
If the content is some Markdown, we don't want to escape it before Markdown (as HTML in markdown is something allowed).

stof replied:
I'm talking abotu the case where `raw=false`. That would apply HTML escaping on the markdown source, before passing it to the `markdown_to_html` filter (because of using a `filter` tag on some output rather than filtering an expression before displaying it)

fabpot replied:
Which means that `raw` should have no effect with Markdown. I'm going to change `markdown()` to take the content as suggested by @lyrixx so that you can use `content()` with `raw` or `markdown()` depending on your needs.

stof replied:
well, the way the template is written right now means that `raw` is **required** if you don't want to have some HTML escaping applied and breaking your markdown.

I suggest changing the template to

```
{{ content|markdown_to_html }}
```

This avoids applying some HTML escaping in a place expecting a markdown context.

fabpot replied:
That's already the case, you haven't looked at the latest version of the PR.

OskarStark commented on src/Symfony/Bridge/Twig/Mime/NotificationEmail.php:
```diff
@@ -0,0 +1,218 @@
+    public const IMPORTANCE_LOW = 'low';
+
+    private const DEFAULT_HTML_TEMPLATE_PATH = '@email/notification.html.twig';
+    private const DEFAULT_TEXT_TEMPLATE_PATH = '@email/notification.txt.twig';
```
Why not adding the default path to the vars itself? The look useless to me

fabpot replied:
Because we are reusing them in the code

fabpot commented:
New iteration: I've replaced the default templates by a theme. One can set a default theme (`setDefaultTheme`, this is configurable globally in the `mailer.yaml` file) and/or set a theme for one email via `theme()`. I've also renamed the default Zurb theme to reflect the change and to allow adding more.

stof commented on composer.json:
```diff
@@ -119,7 +119,10 @@
+        "phpdocumentor/reflection-docblock": "^3.0|^4.0",
+        "twig/cssinliner-extra": "^2.12",
+        "twig/inky-extra": "^2.12",
+        "twig/markdown-extra": "^2.12"
```
wouldn't this forbid testing against Twig 1.x (while at runtime, Symfony will still allow it) ? These extensions don't support Twig 1

fabpot replied:
I know but I don't think there is any other way.

OskarStark replied:
TBH I don’t think it’s worth to drop Twig 1 support and add 3 new dependencies “just” for a notification email.
This is just my POV, please don’t feel offended 🤝

What is the reason not to wait until the ne notifier component gets released?

fabpot replied:
We don't add 3 deps (we are talking about dev requirements here).

I don't see how waiting for the Notifier component will change anything. The 2 features are independant and the code here won't change when the notifier lands.

fabpot replied:
And to be clear, we don't drop Twig 1.x support. We are just not able to run the tests for it in our CI anymore (but as the tests will run in 4.3, I don't think this is a major issue -- and for master, 1.x is dropped anyway).

lyrixx replied:
In a env where we are using twig 1.x we can drop theses 3 deps and skip related tests

fabpot replied:
env: you mean in the CI? Again, I don't  think this is worth it as Twig 1.x is tested on 4.3 and master does not support 1.x.

OskarStark replied:
Sorry I missed that these are dev dependencies. 👍🏻

lyrixx replied:
Yes the CI. 
Anyway I'm okay with your proposal.

yceruto commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -1585,6 +1585,11 @@ private function addMailerSection(ArrayNodeDefinition $rootNode)
                         ->end()
+                        ->arrayNode('notification')
+                            ->children()
+                                ->scalarNode('theme')
```
This configuration may not be necessary if we promote the current Twig override mechanism, e.i:
```yaml
twig:
    paths:
        templates/email: email
```
this would be enough to customize also other templates such as `@email/zurb_2/notification.css` or `'@email/zurb_2/notification_body_markdown.html.twig'`, wdyt?

stof replied:
Changing the default template used for notification emails has a big difference compared to overriding the templates: if you change the template being used, your own template can extend the core template to override some blocks (which is exactly why they have blocks).
Replacing the core templates solves a different use case (and in this case, you need to replace the whole template, as trying to extend would extend yourselves)

yceruto replied:
We have already solved that problem by adding another namespace with the prefix `!`, we could do the same here.

fabpot replied:
Reworked the theme feature based on this feedback (I will update the description accordingly).

OskarStark commented on src/Symfony/Bridge/Twig/Mime/NotificationEmail.php:
```diff
@@ -0,0 +1,226 @@
+        return $this;
+    }
+
+    public static function setDefaultTheme(string $themePath = null)
```
This one is confusing, it sets A theme, but the default theme as default if the value is null? Why not setDefaultTheme without param and a setTheme() with a value and not nullable plus a validation and no fallback to the default theme which is implicit?

fabpot replied:
It sets the default theme globally. Passing `null` means that you want to reset the theme to its default value (I can remove that if it's too confusing). The `setTheme()` method is about changing the theme for that particular instance.

OskarStark replied:
To me calling “setDefaultTheme” should set the default theme, so no parameter needed. 👍🏻👌🏻

fabpot replied:
Fixed as we don't have a default theme anymore to set and `theme()` takes a string as an arg.

fabpot commented:
I've also changed the naming scheme so that we can add more built-in templates over time (`notification` being just the first one).

stof commented on src/Symfony/Bridge/Twig/Resources/views/Email/default:
```diff
@@ -0,0 +1 @@
+zurb_2
```
why is this file ?

stof replied:
if this is a symlink, please test compat with Windows. Symlinks may not work there (using them in recipes is fine, because the source repo for recipes is not cloned by users but by the Flex server, so we don't care about cross-platform support there).

fabpot replied:
This is indeed a symlink. I know that this will probably not work on Windows, not sure how to do otherwise.

stof replied:
well, instead of having a `default` theme being a symlink to zurb_2, why not having a (private) constant in the NotificationEmail to define the name of the default theme (and then loading templates with `zurb_2` in the name) ?

fabpot replied:
Because I want people to be able to override the `default` theme by creating `templates/emails/default/...` files.

stof replied:
well, then I don't see a solution to support Windows if that's the requirement...

yceruto replied:
Why we keep the `zurb_2` directory? could it be named `default` instead?

fabpot replied:
fixed now.

yceruto commented on src/Symfony/Bridge/Twig/Resources/views/Email/zurb_2/notification/body.html.twig:
```diff
@@ -0,0 +1,63 @@
+
+            {% block content %}
+                {% if markdown %}
+                    {{ include('@email/zurb_2/notification/content_markdown.html.twig') }}
```
shouldn't this template be named `@email/default/notification/content_markdown.html.twig` instead?

fabpot replied:
No, we are really in the `zurb_2` theme here.

fabpot commented:
I think  this one is ready now.

yceruto commented:
AppVeyor's failure is related https://ci.appveyor.com/project/fabpot/symfony/builds/27497465#L71

yceruto commented on src/Symfony/Bridge/Twig/Tests/Mime/NotificationEmailTest.php:
```diff
@@ -0,0 +1,66 @@
+    public function testSerialize()
+    {
+        $email = unserialize(serialize((new NotificationEmail())
+            ->markdown('Foo', true)
```
You probably mean `->content()` here? then, the expected raw is `true`.

fabpot replied:
fixed

fabpot commented:
Tests are now as green as possible :)
